include/ddc/kernels/splines/spline_builder_batched.hpp

```diff
@@ -124,16 +128,58 @@ class SplineBuilderBatched
     template <class Layout>
     void operator()(
             ddc::ChunkSpan<double, spline_domain_type, Layout, memory_space> spline,
-            ddc::ChunkSpan<double, vals_domain_type, Layout, memory_space> vals) const;
+            ddc::ChunkSpan<double, vals_domain_type, Layout, memory_space> vals,
+            std::optional<ddc::ChunkSpan<double, derivs_domain_type, Layout, memory_space>> const
+                    derivs_xmin
+            = std::nullopt,
+            std::optional<ddc::ChunkSpan<double, derivs_domain_type, Layout, memory_space>> const
+                    derivs_xmax
+            = std::nullopt) const;
 };
```

Comment:
For DDC I wonder if it would make more sense to provide different operators depending on the boundary conditions? This would mean that the boundary conditions need to be a template parameter but would avoid any overhead associated with `std::optional`

Reply:
That said you would have 4 interfaces possible here so maybe this just gives unnecessary complexity? An optional array is not necessarily much more costly than an array

Reply:
I think this is the kind of very negligeable optimization which degrades readability and maintanability so I don't like it

Reply:
I don't think it degrades maintainability if it's written properly. You just need to have internal functions:
- fill_xmin_derivative_lines_in_rhs
- fill_value_lines_in_rhs(spline, vals)
- fill_xmax_derivative_lines_in_rhs

So then you would have for example (template logic is pseudo-code):
```
    template <class = std::enable_if_t<PERIODIC>>
    void operator()(
        ddc::ChunkSpan<double, spline_domain_type, Layout, memory_space> spline,
        ddc::ChunkSpan<double, vals_domain_type, Layout, memory_space> vals) const {
        fill_value_lines_in_rhs(spline, vals);
    }

    template <class = std::enable_if_t<BcXmin == HERMITE && BcXmax == HERMITE>>
    void operator()(
        ddc::ChunkSpan<double, spline_domain_type, Layout, memory_space> spline,
        ddc::ChunkSpan<double, vals_domain_type, Layout, memory_space> vals) const {
        fill_xmin_derivative_lines_in_rhs(spline, vals);
        fill_value_lines_in_rhs(spline, vals);
        fill_xmax_derivative_lines_in_rhs(spline, vals);
    }
```

You are probably right that this is slightly less readable than the version with `static_assert` but I think it should make the user code more readable as you no longer have to explicitly mention optional in lines like this:
https://github.com/CExA-project/ddc/blob/b0ee2a5cf72d08b5ea22d444f84840f5a930424b/tests/splines/batched_spline_builder.cpp#L309

Reply:
I kind of agree with Emily, it would make sense. Unfortunately i am afraid it could make the usage of it more complicated. I think it would require sfinae `if constexpr` if the user wants to switch boundary conditions depending on the context.

Reply:
wouldn't a dedicated type templated and specialized for the boundary condition make sense?

```
template <Boundary>
class BoundaryParameter;

template <Boundary>
using BoundaryParameter_t = BoundaryParameter<Boundary>::type;

template <Boundary>
using default_boundary_param = BoundaryParameter<Boundary>::default_value;

void operator()(
        ddc::ChunkSpan<double, spline_domain_type, Layout, memory_space> spline,
        ddc::ChunkSpan<double, vals_domain_type, Layout, memory_space> vals,
        BoundaryParameter_t<BcXmin> params_xmin = default_boundary_param<BcXmin>,
        BoundaryParameter_t<BcXmax> params_xmin = default_boundary_param<BcXmax>
    ) const
{
        // [...]
}
```

Reply:
@EmilyBourne I think you also want to handle HERMITE Xmin and GREVILLE Xmax, and vice-versa.

I propose to keep it like that and eventually improve this in an upcoming MR if you think it has to.

Reply:
In an ideal world I think it would be nice to keep this flexibility. I can imagine it being useful for example in a radial direction to use Greville for the centre but Hermite in the wall.
However in practice I have never seen different boundary conditions used at each end of the same spline so I would be okay with a solution which doesn't keep this flexibility

Reply:
I think it requires far more work at this point to remove the flexibility than to keep it ^^, it would means removing BcXmin and BcXmax to put just Bc everywhere. Everything is already working with different boundary conditions so let's continue like that.